src/main.ts

```diff
@@ -100,6 +110,177 @@ export class Minifier {
+    }
+    return null;
+  }
+
+  private _hasAncestor(n: ts.Node, kind: ts.SyntaxKind): boolean {
+    return !!this._getAncestor(n, kind);
+  }
+
+  private _preprocessVisitChildren(node: ts.Node) {
+    node.getChildren().forEach((child) => { this._preprocessVisit(child); });
+  }
+
+  // Key - To: the expected type symbol
+  // Value - From: the actual type symbol
+  // IE: Coercing from type A to type B
+  private _recordCast(from: ts.Symbol, to: ts.Symbol) {
+    if (this._typeCasting.has(from)) {
+      this._typeCasting.get(from).push(to);
+    } else {
+      this._typeCasting.set(from, [to]);
+    }
+  }
+
+  // all preprocess before we start emitting
+  private _preprocessVisit(node: ts.Node) {
+    switch (node.kind) {
+      case ts.SyntaxKind.CallExpression: {
+        var callExpr = <ts.CallExpression>node;
+        var lhsSymbol = this._typeChecker.getSymbolAtLocation(callExpr.expression);
+
+        let paramSymbols: ts.Symbol[] = [];
+
+        // TODO: understand cases of multiple declarations, pick first declaration for now
+        if (!lhsSymbol || !((<any>lhsSymbol.declarations[0]).parameters)) {
+          this._preprocessVisitChildren(node);
+          break;
+        } else {
+          (<any>lhsSymbol.declarations[0])
+              .parameters.forEach((param) => { paramSymbols.push(param.type.symbol); });
+
+          let argsSymbols: ts.Symbol[] = [];
+
+          // right hand side argument has actual type of parameter
+          callExpr.arguments.forEach(
+              (arg) => { argsSymbols.push(this._typeChecker.getTypeAtLocation(arg).symbol); });
+
+          // Casting from: Use site symbol, to: actual parameter type
+          paramSymbols.forEach((sym, i) => { this._recordCast(sym, argsSymbols[i]); });
+
+          this._preprocessVisitChildren(node);
+          break;
+        }
+      }
+      case ts.SyntaxKind.VariableDeclaration: {
+        let varDecl = <ts.VariableDeclaration>node;
+        if (varDecl.initializer && varDecl.type) {
+          let varDeclTypeSymbol = this._typeChecker.getTypeAtLocation(varDecl.type).symbol;
+          let initTypeSymbol = this._typeChecker.getTypeAtLocation(varDecl.initializer).symbol;
+
+          // Casting from: initializer's type symbol, to: actual variable declaration's annotated
+          // type
+          this._recordCast(initTypeSymbol, varDeclTypeSymbol);
+        }
+        this._preprocessVisitChildren(node);
+        break;
+      }
+      case ts.SyntaxKind.ReturnStatement: {
+        // check if there is an expression on the return statement since it's optional
+        if (node.parent.kind !== ts.SyntaxKind.SourceFile &&
+            (<ts.ReturnStatement>node).expression) {
+          let symbolReturn =
+              this._typeChecker.getTypeAtLocation((<ts.ReturnStatement>node).expression).symbol;
+          let ancestor;
+
+          let hasMethodDeclAncestor = this._hasAncestor(node, ts.SyntaxKind.MethodDeclaration);
```

Comment:
this is very inefficient, I would suggest using _getAncestor and using `null` value as a check that the ancestor does not exist.